pkg/services/ngalert/state/persister_sync.go

```diff
@@ -0,0 +1,111 @@
+package state
+
+import (
+	"context"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/grafana/dskit/concurrency"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	ngModels "github.com/grafana/grafana/pkg/services/ngalert/models"
+)
+
+type SyncStatePersister struct {
+	log   log.Logger
+	store InstanceStore
+	// doNotSaveNormalState controls whether eval.Normal state is persisted to the database and returned by get methods.
+	doNotSaveNormalState bool
+	// maxStateSaveConcurrency controls the number of goroutines (per rule) that can save alert state in parallel.
+	maxStateSaveConcurrency int
+}
+
+func NewSyncStatePerisiter(log log.Logger, store InstanceStore, doNotSaveNormalState bool, maxStateSaveConcurrency int) StatePersister {
+	return &SyncStatePersister{
+		log:                     log,
+		store:                   store,
+		doNotSaveNormalState:    doNotSaveNormalState,
+		maxStateSaveConcurrency: maxStateSaveConcurrency,
+	}
+}
+
+func (a *SyncStatePersister) Async(_ context.Context, _ *clock.Ticker, _ *cache) {
```

Comment:
This suggests to me the design of the interface needs to be revisited. We should be able to pass different implementations of the interface without also having to change the behaviour in the caller.

The main issue here is that the interface is unsafe against accidental no-ops. In addition, functions that accept this interface need to work out what the concrete type is in order to call the right method. But having an additional boolean (feature flag) outside the interface is not the answer either.

I proposed a solution to this problem in [here](https://github.com/grafana/grafana/pull/74998#discussion_r1334966658). Would that not be better?

Reply:
We discussed this extensively in the first PR and offline in the past. You will always have this, as both are semantically very different, one running in the foreground, needing an invocation through a function call - and the other running in the background not needing any invocation and just reading from the cache.

One thing we could have done is extract the cache and make it public so we can directly pass it to the async constructor, not needing the async function in the interface and just having a `Persist` one that does something in the sync implementation and nothing in the async.

But having in the back of our head that this will become the new default and that we remove the sync eventually, I don't think this is as bad and would justify the effort described above.